### DISCUSSION TITLE
How to show CJK characters in button label or window title?

### DISCUSSION THREAD

**Discussion by thelastfantasy** [gtk-rs/gtk4-rs#1230] [Q&A]
Hello I'm learning gtk4-rs with https://gtk-rs.org/gtk4-rs/stable/latest/book/hello_world.html
I changed the label to include some Chinese characters in the sample code but got this error message.

I think there would be some problems with the encoding but I don't know how to solve it.
Would anyone please help me?

>    Compiling my-gtk-app v0.1.0 (D:\Dev\my-gtk-app)
error: linking with `link.exe` failed: exit code: 1104                                                                                                                                                                                             
  |
  = note: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.30.30705\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "C:\\Users\\TOGYOK~1\\AppData\\Local\\Temp\\rustcSDSf2s\\symbols.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.1co10xg1kb5mb0t0.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.1cq8z0myja6erekf.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.1exlsjf8i29jlxjo.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.1f6pnim33tda4k1q.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.1sfl43a9gyafghqy.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.20s42nkhzklldu1c.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.22gfutru3uikm5k4.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.26uat6mngvnjr18j.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.2b7eyr5l2b1jivin.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.2iwf54ohb6gwmn8o.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.2l6j0lv0wqji6jyy.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.2oe20ezhqsc0zepl.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.2zcg8912tqoqf9yy.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.3096hkccsjkh0v3p.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.31j59sspeikmu3nd.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.36j5kp1zdq8nhpsp.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.3b8y4zu1w2uzg93m.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.3blcxosqhjal2toa.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.3cvi0s3j6bgc39es.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.3fy9ixh0qnu99rh2.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.3jrv8m9hnxihi3t5.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.3ledjawm35soi58s.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.3zj9sara24bzji0c.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.47262vwvbnvaloip.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.4h7lyl1832cucy9q.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.4low3x7lkvqof715.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.4rc6rpjwv77o6801.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.52vugrkp414hywxv.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.56lo42lzb08v75qc.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.6fs1hk5yp6es3gm.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.dhv8xhrt29vzb2y.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.s0ozn4mtv5bzovq.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.sa4enr5n5v57nn1.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.sj7ml1785munwpt.rcgu.o" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.3gxev7hfdkpbd279.rcgu.o" "/LIBPATH:D:\\Dev\\my-gtk-app\\target\\debug\\deps" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:/gnome/lib" "/LIBPATH:C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgtk4-c09378135771a61e.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgsk4-ed81fdf084a48e21.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgraphene-f29026af3866b2c9.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgdk4-1b1cd90abd8d31bd.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libpango-48a6ddd519bf70f5.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgdk_pixbuf-0169a5949ddf97e3.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgio-f54ff59ee62a46fd.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libfutures_io-2cb7e2a5e9d9954d.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libcairo-b9128ca0be33eb68.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libglib-e43bd0614235ba43.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libfutures_executor-04599cac272186dc.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libsmallvec-4a23ea4a75e7ae04.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libthiserror-e9586a11a56e98a2.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libfutures_util-572cc8bc5a68cbed.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libslab-853c1231bb14a9a9.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libpin_project_lite-232ef7cd835419c5.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libpin_utils-412ae58bb4eab159.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libfutures_channel-22b87bc1f1636708.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libfutures_task-d1d35dae856c2c14.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libfutures_core-3152cc7331141c3c.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libonce_cell-c4d57e5041e1b607.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libbitflags-923ab9399e3495d4.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgtk4_sys-5f2952ed7ad6083b.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgsk4_sys-ebfec89edcfc8041.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgraphene_sys-2f3130155711b96a.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgdk4_sys-36de16efaf7ad3e9.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgdk_pixbuf_sys-1aba2545a50f111c.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libpango_sys-b0b81f3a828b8755.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libcairo_sys-ffeba07edee7573a.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgio_sys-95ac84f3ee988113.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libwinapi-abefa8d9104ad9db.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libgobject_sys-dd1c5996e131e2cf.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libglib_sys-6261680f71da3179.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\liblibc-5d0a97a89ec6f6e9.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libfield_offset-3ca88e32296e1e66.rlib" "D:\\Dev\\my-gtk-app\\target\\debug\\deps\\libmemoffset-3606c0f24cdb97cc.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd-fdaa63d6970be740.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libpanic_unwind-a684236a9da37ec6.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_demangle-85c151aafb35bf5c.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd_detect-7dbe75e48c9029c0.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libhashbrown-78044ac8790346aa.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libminiz_oxide-8743285c70baade8.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libadler-b260ead09cc3ca75.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-200564555b6c364f.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libunwind-60e4ae6d068e334f.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-d74559c1731cb9bf.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liblibc-3d51e2979c44c27c.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liballoc-268ac0a196c81649.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_core-e74895db82cfb265.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcore-d805a250e2b7e197.rlib" "C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-3f507ba7adac1c2a.rlib" "gobject-2.0.lib" "gtk-4.lib" "pangowin32-1.0.lib" "pangocairo-1.0.lib" "pango-1.0.lib" "gdk_pixbuf-2.0.lib" "cairo-gobject.lib" "cairo.lib" "graphene-1.0.lib" "gio-2.0.lib" "gobject-2.0.lib" "glib-2.0.lib" "intl.lib" "gtk-4.lib" "pangowin32-1.0.lib" "pangocairo-1.0.lib" "pango-1.0.lib" "gdk_pixbuf-2.0.lib" "cairo-gobject.lib" "cairo.lib" "graphene-1.0.lib" "gio-2.0.lib" "gobject-2.0.lib" "glib-2.0.lib" "intl.lib" "graphene-1.0.lib" "gobject-2.0.lib" "glib-2.0.lib" "intl.lib" "gtk-4.lib" "pangowin32-1.0.lib" "pangocairo-1.0.lib" "pango-1.0.lib" "gdk_pixbuf-2.0.lib" "cairo-gobject.lib" "cairo.lib" 
"graphene-1.0.lib" "gio-2.0.lib" "gobject-2.0.lib" "glib-2.0.lib" "intl.lib" "gdk_pixbuf-2.0.lib" "gobject-2.0.lib" "glib-2.0.lib" "intl.lib" "pango-1.0.lib" "gobject-2.0.lib" "glib-2.0.lib" "intl.lib" "cairo-gobject.lib" "gobject-2.0.lib" "glib-2.0.lib" "intl.lib" "cairo.lib" "gio-2.0.lib" "gobject-2.0.lib" "glib-2.0.lib" "intl.lib" "advapi32.lib" "cfgmgr32.lib" "kernel32.lib" "user32.lib" "ws2_32.lib" "gobject-2.0.lib" "glib-2.0.lib" "intl.lib" "gobject-2.0.lib" "glib-2.0.lib" "intl.lib" "legacy_stdio_definitions.lib" "kernel32.lib" "advapi32.lib" "userenv.lib" "kernel32.lib" "ws2_32.lib" "bcrypt.lib" "msvcrt.lib" "legacy_stdio_definitions.lib" "/NXCOMPAT" "/LIBPATH:C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "/OUT:D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.exe" "/OPT:REF,NOICF" "/DEBUG" "/NATVIS:C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:C:\\Users\\To Gyokuryu\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libstd.natvis"
  = note: Non-UTF-8 output: LINK : fatal error LNK1104: \xce\xde\xb7\xa8\xb4\xf2\xbf\xaa\xce\xc4\xbc\xfe\xa1\xb0D:\\Dev\\my-gtk-app\\target\\debug\\deps\\my_gtk_app.exe\xa1\xb1\r\n

**Comment by sdroege** [comment]
Rust source files need to be UTF-8 but it looks from the error message like you added the Chinese characters in a different encoding. Can you check that?

**Comment by thelastfantasy** [comment]
Thank you for your reply!
I think it's because I forgot to close the "my-gtk-app.exe" instance.
my-gtk-app.exe file can be launched from explorer after `cargo build` now.

But another problem is, my-gtk-app.exe executable file can only run from explorer (double click), when I tried to run it with `cargo run` I always got this error↓

> error: process didn't exit successfully: `target\debug\my-gtk-app.exe` (exit code: 0xc0000139, STATUS_ENTRYPOINT_NOT_FOUND)

Maybe I need to create a build.rs file?

**Reply by sdroege** [reply]
You need to make sure that all required DLLs are in the `PATH` when running the application, as always on Windows. You can use https://github.com/lucasg/Dependencies to list all dependencies of your executable and to get an idea which ones are not found.
